implementations/rust/ockam/ockam_command/src/terminal/mod.rs

```diff
@@ -347,6 +349,29 @@ impl<W: TerminalWriter> Terminal<W, ToStdErr> {
         Ok(output)
     }
 
+    pub fn build_json_list(&self, nodes: &[impl Serialize], empty_message: &str) -> Result<String> {
+        if nodes.is_empty() {
+            let output = format!("{}", &fmt_warn!("{empty_message}"));
+            return Ok(output);
+        }
+
+        let mut serialized_nodes = Vec::with_capacity(nodes.len());
+
+        for node in nodes {
+            let json_node = serde_json::to_string_pretty(node).map_err(|e| {
+                Error::InternalError {
+                    error_message: e.to_string(),
+                    exit_code: exitcode::SOFTWARE
+                }
+            })?;
+            serialized_nodes.push(json_node);
+        }
+
+        let output = serialized_nodes.join(",\n");
```

Comment:
The serialization should be done in one step to avoid any manual intervention with the serialization process. Since the input objects implement the `Serialize` trait, you could just serialize the whole input array as is.

```
pub fn build_json_list(&self, items: &[impl Serialize], empty_message: &str) -> Result<String> {
    if items.is_empty() {
        let output = format!("{}", &fmt_warn!("{empty_message}"));
        return Ok(output);
    }
    Ok(serde_json::to_string_pretty(items)?)
}
```


Furthermore, note that the `Terminal` instance knows about the output format already, so you don't need to do the check externally. Rather the `build_list` method should check the format and print the items in the expected format. You might need to update the bounds requirements of the `items` argument to include the `Serialize` trait:

```
pub fn build_list(
    &self,
    items: &Vec<impl crate::output::Output>,
    header: &str,
    empty_message: &str,
) -> Result<String> {
    match self.output_format {
        OutputFormat::Plain => {
            ..
        }
        OutputFormat::Json => {
            ..
        }
    }
)
```

Reply:
Thanks for the great input @adrianbenavides. The two I think we should consider before putting everything in `build_list` is:

1. There are a lot of types that are passed into `build_list` that do not implement `Serialize`. `CredentialOutput`, `IdentityListOutput`, and `SecureChannelListOutput`, just to name a few. So adding `Serialize` as an additional trait bound would require that I derive `Serialize` for all of those types. I'm happy to do that if you think that's appropriate.

2. The `header` argument isn't necessary it seems for the JSON output, so passing`header` into `build_list` when the output format is JSON seems a bit awkward, but it feels like a minor issue.

So all-in-all I agree with your suggestions — just want your approval before I start deriving `Serialize` for everything that is passed into `build_list`.

Reply:
I was reviewing the code around building the output and I think we have a better way to do this. Since generating the json format is as easy as calling `serde_json::to_string_pretty(items)?`, we could leave the `build_list` as is and do the following:

```rust
let plain = opts
    .terminal
    .build_list(&nodes, "Nodes", "No nodes found on this system.")?;
let json = serde_json::to_string_pretty(&nodes).into_diagnostic()?;
opts.terminal
    .stdout()
    .plain(plain)
    .json(json)
    .write_line()?;
```

That way you don't have to worry about other commands. We can take care of the rest in separate PR's.